### PR TITLE
✨ start a new package

### PR DESCRIPTION
**Pull request by michenly**
## Description

Fixes (issue #) https://github.com/Shopify/quilt/issues/1067

Unifying these 3
1. https://github.com/Shopify/web/blob/master/packages/%40shopify/koa-performance/middleware.ts#L66
2. https://github.com/Shopify/quilt/blob/master/packages/koa-metrics/src/Metrics.ts#L44
3. https://github.com/Shopify/web/blob/master/packages/%40shopify/statsd/client.ts#L10

## 🎩 
1. Using this companion branch in web
https://github.com/Shopify/web/pull/18952
2. start by creating folder for the new package in `web/node_modules/@shopify/statsd` and copy the `package.json` over
3. `yarn tophat statsd ../web` in quilt

## Todo

- [x] Bring in all the `snakeCase ` customization from https://github.com/Shopify/web/blob/master/packages/%40shopify/statsd/client.ts
- [x] Unify with https://github.com/Shopify/quilt/blob/master/packages/koa-metrics/src/Metrics.ts#L91
- [x] add test to the client
- [x] add documentation to the package

### CODE REVIEW

File: packages/koa-metrics/package.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
   },
   "homepage": "https://github.com/shopify/quilt/blob/master/packages/koa-metrics/README.md",
   "dependencies": {
+    "@shopify/statsd": "^1.0.0",
```
Comment (michenly) [inline packages/koa-metrics/package.json]: can I do this yet? I remember us discussing a fix recently. 
Or would I need to wait till `@shopify/statsd` is actually publish.

File: packages/koa-metrics/src/Metrics.ts [hunk 2]
```diff
@@ -1,104 +0,0 @@
...
-      host,
-      port,
-      prefix,
-      globalTags,
-    });
-    this.client = this.rootClient;
-  }
-
-  public distribution(name: string, value: number, tags?: TagsMap) {
-    this.log(MetricType.Distribution, name, value, tags);
```
Comment (michenly) [inline packages/koa-metrics/src/Metrics.ts]: this is the only line I did not carry over

**Comment by michenly** [comment]
Still adding test & document. But the implementation can be review + 🎩 now!

File: packages/koa-metrics/package.json (hunk 1 above)
Comment (michenly) [reply packages/koa-metrics/package.json]: ya I can't...grr...

File: README.md [hunk 3]
```diff
@@ -96,10 +97,10 @@ Check out our [Contributing Guide](./.github/CONTRIBUTING.md)
 
 ## Questions?
 
-For Shopifolk, you can reach out to us in Slack in the `#web-foundation-tech` channel. For external inquiries, we welcome bug reports, enhancements, and feature requests via Github issues.
+For Shopifolk, you can reach out to us in Slack in `#web-foundation-tech` channel. For external inquiries, we welcome bug reports, enhancements, and feature requests via Github issues.
```
Comment (TheMallen) [inline README.md]: Either remove `channel` or add back the `the` here please :)

File: packages/statsd/src/client.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+export interface Options extends ClientOptions {
+  logger?: Logger;
+  snakeCase?: boolean;
+}
+
+export class StatsDClient {
+  private statsd: StatsD;
+  private logger: Logger | null = null;
```
Comment (TheMallen) [inline packages/statsd/src/client.ts]: can we default this to a dirt simple object that just exposes console.log when there is no logger provided?

File: packages/statsd/src/client.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  private options: Options;
+
+  constructor({logger, ...options}: Options) {
+    this.logger = logger || null;
+    this.options = {
+      ...options,
+      errorHandler: options.errorHandler
+        ? options.errorHandler
+        : this.handleError.bind(this),
```
Comment (TheMallen) [inline packages/statsd/src/client.ts]: Can we just make it an error function and not bother with this bind?

File: packages/statsd/src/client.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  ) {
+    return (error?: Error) => {
+      if (error) {
+        if (this.options.errorHandler) {
+          this.options.errorHandler(error);
+        }
+        reject(error);
+      }
+
+      // Failing to send the metrics is not something we need to propagate
```
Comment (TheMallen) [inline packages/statsd/src/client.ts]: This comment is a bit of a lie, as far as I can tell from the way this behaves / the code in general if we get here it is because there was no error (therefore sending stats succeeded). We can just🔥 this comment.

File: packages/statsd/src/client.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    if (this.logger == null) {
+      return;
+    }
+
+    this.logger.log(
+      `Error occurred in the StatsD client:\n${error.stack || error.message}`,
+    );
+  }
+
+  private createErrorCallback<T>(
```
Comment (TheMallen) [inline packages/statsd/src/client.ts]: `errorCallback` is a bit misleading here, since the returned function will get called on every metric regardless of pass or fail

File: packages/statsd/src/test/client.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+describe('statsd', () => {
+  it.todo('works');
```
Comment (TheMallen) [inline packages/statsd/src/test/client.ts]: I'd like to see some tests here, not sure exactly how we could make them meaningful. Looking at `hot-shots` itself they setup a weird local server to see that it receives the expected requests. :/

I'm ok with them coming in as a subsequent PR at worst.

File: packages/statsd/README.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+An opinionated StatsD client for Shopify Node.js server and other StatsD utilities.
+
+## Installation
+
+```bash
+$ yarn add @shopify/statsd
+```
+
+## Usage
```
Comment (TheMallen) [inline packages/statsd/README.md]: We should add these before merging :)

File: packages/statsd/src/test/client.test.ts [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+      const error = new Error('Something went wrong!');
+      statsDMock.close.mockImplementation(callback => {
+        callback(error);
+      });
+
+      statsDClient.close();
+      expect(errorHandler).toHaveBeenCalledWith(error);
+    });
+
+    it("calls logger's log on the promise when the client returns an error and there is no errorHandler in options", () => {
```
Comment (TheMallen) [inline packages/statsd/src/test/client.test.ts]: Personally I'd just🔥 the tests that test what we log since we generally don't test logs in other libraries.

File: packages/statsd/src/client.ts [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    });
+  }
+
+  addGlobalTags(globalTags: Tags) {
+    this.statsd = this.statsd.childClient({
+      globalTags,
+    });
+  }
+
+  initTimer(): Timer {
```
Comment (TheMallen) [inline packages/statsd/src/client.ts]: I'm not super sure this makes sense as part of this class tbh, it seems unrelated to the rest, and doesn't share any state.

Comment (michenly) [reply packages/statsd/src/client.ts]: good call, 🔥ing
